contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3661,101 +3662,151 @@ get_insert_bulk_kilobytes_per_batch()
+		if ((strcmp(rolname, user) == 0) || pg_namespace_ownercheck(schemaOid, role_oid) || is_member_of_role(role_oid, datdba))
+			ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					errmsg("Cannot grant, deny, or revoke permissions to sa, dbo, entity owner, information_schema, sys, or yourself.")));
 
-			parsetree_list = gen_grantschema_subcmds(schema_name, rolname, stmt->is_grant, stmt->with_grant_option, priv_name);
-			/* Run all subcommands */
-			foreach(parsetree_item, parsetree_list)
-			{
-				Node	   *stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
-				PlannedStmt *wrapper;
-
-				/* need to make a wrapper PlannedStmt */
-				wrapper = makeNode(PlannedStmt);
-				wrapper->commandType = CMD_UTILITY;
-				wrapper->canSetTag = false;
-				wrapper->utilityStmt = stmt;
-				wrapper->stmt_location = 0;
-				wrapper->stmt_len = 0;
-
-				/* do this step */
-				ProcessUtility(wrapper,
-							"(GRANT SCHEMA )",
-							false,
-							PROCESS_UTILITY_SUBCOMMAND,
-							NULL,
-							NULL,
-							None_Receiver,
-							NULL);
-
-				/* make sure later steps can see the object created here */
-				CommandCounterIncrement();
-			}
-			/* Add entry for each grant statement. */
-			if (stmt->is_grant && !check_bbf_schema_for_entry(stmt->schema_name, "ALL", priv_name, rolname))
-				add_entry_to_bbf_schema(stmt->schema_name, "ALL", priv_name, rolname, NULL);
-			/* Remove entry for each revoke statement. */
-			if (!stmt->is_grant && check_bbf_schema_for_entry(stmt->schema_name, "ALL", priv_name, rolname))
-			{
-				/* If any object in the schema has the OBJECT level permission. Then, internally grant that permission back. */
-				grant_perms_to_objects_in_schema(stmt->schema_name, priv_name, rolname);
-				del_from_bbf_schema(stmt->schema_name, "ALL", priv_name, rolname);
-			}
+		if (!is_member_of_role(GetSessionUserId(), datdba) && !login_is_db_owner && !pg_namespace_ownercheck(schemaOid, GetUserId()))
+			ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("Cannot find the schema \"%s\", because it does not exist or you do not have permission.", stmt->schema_name)));
+
+		if((privilege_maskInt & 32) == 32)
+			exec_gen_grantschema(schema_name, rolname, stmt, "execute");
+		if((privilege_maskInt & 16) == 16)
+			exec_gen_grantschema(schema_name, rolname, stmt, "select");
+		if((privilege_maskInt & 8) == 8)
+			exec_gen_grantschema(schema_name, rolname, stmt, "insert");
+		if((privilege_maskInt & 4) == 4)
+			exec_gen_grantschema(schema_name, rolname, stmt, "update");
+		if((privilege_maskInt & 2) == 2)
+			exec_gen_grantschema(schema_name, rolname, stmt, "delete");
+		if((privilege_maskInt & 1) == 1)
+			exec_gen_grantschema(schema_name, rolname, stmt, "references");
+
+		/* Add entry for each grant statement. */
+		if (stmt->is_grant && !check_bbf_schema_for_entry(stmt->schema_name, "ALL", rolname))
```

Comment:
Better to store the result of check_bbf_schema_for_entry if you need to call it multiple times

Reply:
We are calling this function with different arguments, storing the result is not needed.

Reply:
I mean the repetitive check in the if else here. Is schema_name or rolname changing?